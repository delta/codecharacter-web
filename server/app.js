const express = require('express');
const path = require('path');
const app = express();

app.use(express.static(path.join(__dirname, '..', 'build')));

app.get('/*', (req, res) => {
  res.sendFile(path.join(__dirname, '..', 'build', 'index.html'));
});

app.listen(5000, '0.0.0.0');
console.log('Server is listening on http://localhost:5000');
