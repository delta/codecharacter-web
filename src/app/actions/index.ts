export * from 'app/actions/Dashboard';
export * from 'app/actions/Leaderboard';
export * from 'app/actions/code/EditorSettings';
export * from 'app/actions/GameLog';
export * from 'app/actions/code/Code';
export * from 'app/actions/User';
export * from 'app/actions/Notification';
export * from 'app/actions/code/Submission';
export * from 'app/actions/MatchView';
export * from 'app/actions/ProfileUser';
