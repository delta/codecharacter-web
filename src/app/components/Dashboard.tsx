import { ErrorBoundary } from 'app/components/ErrorBoundary';
import ReactTour from 'app/components/ReactTour';
import CodeStatus from 'app/containers/code/CodeStatus';
import Editor from 'app/containers/code/Editor';
import GameLog from 'app/containers/GameLog';
import Renderer from 'app/containers/Renderer';
import SideBar from 'app/containers/SideBar';
import SidePanel from 'app/containers/SidePanel';
import SocketHandler from 'app/containers/SocketHandler';
import SubmitBar from 'app/containers/SubmitBar';
import { Routes } from 'app/routes';
import * as style from 'app/styles/Dashboard.css';
import * as DashboardInterfaces from 'app/types/Dashboard';
import * as React from 'react';
import { Grid, Row } from 'react-bootstrap';
import * as ReactGA from 'react-ga';
import { Redirect } from 'react-router-dom';
import { GOOGLE_ANALYTICS_TRACKING_ID } from '../../config/config';

/* tslint:disable-next-line:import-name */
import SplitPane from 'react-split-pane';
import { Welcome } from './Welcome';

export class Dashboard extends React.Component<
  DashboardInterfaces.Props,
  DashboardInterfaces.State
> {
  public sideBarWidth = 50;
  public sidePanelWidth = 380;
  public minEditorWidth = 350;
  public minRendererWidth = 50;
  public initialEditorRatio = 0.4;
  public initialRendererHeight = 400;
  public compilationData = '';

  constructor(props: DashboardInterfaces.Props) {
    super(props);

    const fixedLeftPaneWidth =
      this.sideBarWidth + (this.props.sidePanelOpen ? this.sidePanelWidth : 0);

    this.state = {
      fixedLeftPaneWidth,
      editorWidthRatio: this.initialEditorRatio,
      isReactTourActive: false,
      rendererHeight: this.initialRendererHeight,
      splitPaneState: DashboardInterfaces.SplitPaneState.BOTH,
      windowWidth: window.innerWidth,
    };

    if (window.addEventListener) {
      window.addEventListener('resize', this.onWindowResize, true);
    }
  }

  public componentDidMount() {
    ReactGA.initialize(GOOGLE_ANALYTICS_TRACKING_ID);
    ReactGA.pageview('/');
  }

  public componentWillReceiveProps(nextProps: DashboardInterfaces.Props) {
    const { sidePanelOpen } = nextProps;
    sidePanelOpen ? this.onToggleSidePanel(true) : this.onToggleSidePanel(false);
  }

  public render() {
    const {
      editorWidthRatio,
      windowWidth,
      fixedLeftPaneWidth,
      splitPaneState,
      isReactTourActive,
    } = this.state;
    const {
      isLoggedIn,
      isAuthenticationOpen,
      setIsAuthenticationOpen,
      isWelcomeModalOpen,
      closeWelcomeModal,
    } = this.props;

    let editorWidth;
    switch (splitPaneState) {
      case DashboardInterfaces.SplitPaneState.EDITOR: {
        editorWidth = windowWidth - fixedLeftPaneWidth - this.minRendererWidth;
        editorWidth = Math.max(editorWidth, this.minEditorWidth);
        break;
      }
      case DashboardInterfaces.SplitPaneState.BOTH: {
        const availableWidth = windowWidth - fixedLeftPaneWidth;
        editorWidth = editorWidthRatio * availableWidth;
        editorWidth = Math.max(editorWidth, this.minEditorWidth);
        break;
      }
      case DashboardInterfaces.SplitPaneState.RENDERER: {
        editorWidth = 0;
        break;
      }
      default: {
        editorWidth = 0;
      }
    }
    if (!isLoggedIn) {
      return <Redirect to={Routes.LOGIN} />;
    }

    return (
      <div>
        {isWelcomeModalOpen ? <Welcome closeWelcomeModal={() => closeWelcomeModal()} /> : null}
        {isLoggedIn && isReactTourActive && !isWelcomeModalOpen ? (
          <ReactTour toggleReactTour={this.onToggleReactTour} />
        ) : null}
        {isLoggedIn ? <SocketHandler /> : null}
        <SplitPane
          style={{
            filter: `blur(${isAuthenticationOpen || isWelcomeModalOpen ? 2 : 0}px)`,
            transition: 'filter 0.7s',
          }}
          split="vertical"
          minSize={this.minEditorWidth + fixedLeftPaneWidth}
          maxSize={windowWidth - this.minRendererWidth}
          size={this.state.fixedLeftPaneWidth + editorWidth}
          resizerClassName={style.vertical}
          onChange={this.onSplitPaneResize}
          allowResize={splitPaneState === DashboardInterfaces.SplitPaneState.BOTH}
        >
          <Grid fluid={true}>
            <Row>
              <div
                style={{
                  height: '100vh',
                }}
              >
                <SideBar
                  toggleReactTour={this.onToggleReactTour}
                  setIsAuthenticationOpen={setIsAuthenticationOpen}
                />
              </div>
              <SidePanel sidePanelWidth={this.sidePanelWidth} />
              <div
                style={{
                  height: '100vh',
                }}
              >
                <CodeStatus width={editorWidth} />
                {this.state.splitPaneState !== DashboardInterfaces.SplitPaneState.RENDERER ? (
                  <div id="editor_div">
                    <Editor editorWidth={editorWidth} />
                  </div>
                ) : null}
              </div>
            </Row>
          </Grid>
          <SplitPane
            split="horizontal"
            defaultSize={this.initialRendererHeight}
            resizerClassName={style.horizontal}
            onChange={(size: number) => this.setState({ rendererHeight: size })}
            maxSize={500}
          >
            <Grid fluid={true} style={{ backgroundColor: '#141414' }}>
              <Row>
                <SubmitBar
                  changeSplitPaneState={this.changeSplitPaneState}
                  splitPaneState={this.state.splitPaneState}
                />
              </Row>
              <Row style={{ paddingTop: 42 }} id="renderer">
                <ErrorBoundary>
                  <Renderer height={this.state.rendererHeight - 42} />
                </ErrorBoundary>
              </Row>
            </Grid>
            <GameLog height={window.innerHeight - this.state.rendererHeight} />
          </SplitPane>
        </SplitPane>
      </div>
    );
  }

  private onWindowResize = () => {
    this.setState({
      windowWidth: window.innerWidth,
    });
  };

  private changeSplitPaneState = (state: DashboardInterfaces.SplitPaneState): void => {
    this.setState({
      splitPaneState: state,
    });
  };

  private onSplitPaneResize = (leftPaneSize: number): void => {
    const availableWidth = this.state.windowWidth - this.state.fixedLeftPaneWidth;
    let editorWidth = leftPaneSize - this.state.fixedLeftPaneWidth;
    editorWidth = Math.max(editorWidth, this.minEditorWidth);
    const editorWidthRatio = editorWidth / availableWidth;

    if (this.state.splitPaneState === DashboardInterfaces.SplitPaneState.BOTH) {
      this.setState({
        editorWidthRatio,
      });
    }
  };

  private onToggleSidePanel = (isSidePanelOpen: boolean): void => {
    const fixedLeftPaneWidth = this.sideBarWidth + (isSidePanelOpen ? this.sidePanelWidth : 0);

    this.setState({
      fixedLeftPaneWidth,
    });
  };

  private onToggleReactTour = (): void => {
    this.setState({
      isReactTourActive: !this.state.isReactTourActive,
    });
  };
}
