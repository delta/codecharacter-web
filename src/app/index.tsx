import ActivateUser from 'app/containers/Authentication/ActivateUser';
import Login from 'app/containers/Authentication/Login';
import Register from 'app/containers/Authentication/Register';
import Dashboard from 'app/containers/Dashboard';
import LandingPage from 'app/containers/LandingPage';
import Leaderboard from 'app/containers/Leaderboard';
import UserProfileModal from 'app/containers/UserProfileModal';
import { Routes } from 'app/routes';
// @ts-ignore
import { initializeRendererAssets } from 'code-character-renderer-20';
import * as React from 'react';
import { hot } from 'react-hot-loader';
// @ts-ignore
import { BrowserRouter, Route, Switch } from 'react-router-dom';

initializeRendererAssets();

/* tslint:disable-next-line:variable-name */
export const App = hot(module)(() => (
  <BrowserRouter>
    <Switch>
      <Route exact path={Routes.ROOT} component={Dashboard} />
      <Route exact path={Routes.LOGIN} component={Login} />
      <Route exact path={Routes.REGISTER} component={Register} />
      <Route exact path={Routes.LEADERBOARD} component={Leaderboard} />
      <Route exact path={Routes.USER_PROFILE_MODEL} component={UserProfileModal} />
      <Route path={Routes.USER_ACTIVATION} component={ActivateUser} />
      <Route exact path={Routes.HOME} component={LandingPage} />
    </Switch>
  </BrowserRouter>
));
