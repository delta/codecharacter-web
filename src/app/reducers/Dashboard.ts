import { DashboardActions } from 'app/actions';
import * as DashboardInterfaces from 'app/types/Dashboard';

export enum SidePanelTab {
  DOCS = 'DOCS',
  EDITOR_SETTINGS = 'EDITOR_SETTINGS',
  LEADERBOARD = 'LEADERBOARD',
  COMMIT_LOG = 'COMMIT_LOG',
  MATCH = 'MATCH',
  NOTIFICATION = 'NOTIFICATION',
  NONE = 'NONE',
}

const dashboardStoreIntialState: DashboardInterfaces.DashboardStoreState = {
  isWelcomeModalOpen: false,
  sidePanelTab: SidePanelTab.NONE,
};

export const dashboardReducer = (
  state = dashboardStoreIntialState,
  action: DashboardInterfaces.DashboardStoreAction,
) => {
  switch (action.type) {
    case DashboardActions.Type.SET_SIDE_PANEL_TAB:
      return {
        ...state,
        sidePanelTab: action.payload.sidePanelTab,
      };
    case DashboardActions.Type.RESET_DASHBOARD_STATE: {
      return {
        ...dashboardStoreIntialState,
      };
    }
    case DashboardActions.Type.SET_IS_WELCOME_MODAL_OPEN: {
      return {
        ...state,
        isWelcomeModalOpen: action.payload.isOpen,
      };
    }
    default:
      return state;
  }
};
