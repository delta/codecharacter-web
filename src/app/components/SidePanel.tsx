import { Docs } from 'app/components/Docs';
import CommitLog from 'app/containers/code/CommitLog';
import EditorSettings from 'app/containers/code/EditorSettings';
import Leaderboard from 'app/containers/Leaderboard';
// tslint:disable-next-line:import-name
import Match from 'app/containers/MatchView';
import Notification from 'app/containers/Notification';
import { SidePanelTab } from 'app/reducers/Dashboard';
import * as styles from 'app/styles/SidePanel.module.css';
import classnames from 'classnames';
import * as React from 'react';

export class SidePanel extends React.Component<SidePanel.Props, {}> {
  public render() {
    const { sidePanelTab, sidePanelWidth } = this.props;
    let sidePanel = <div className={classnames(styles.SidePanel)} />;
    switch (sidePanelTab) {
      case SidePanelTab.NONE:
        sidePanel = <div className={classnames(styles.SidePanel)} />;
        break;
      case SidePanelTab.EDITOR_SETTINGS:
        sidePanel = <EditorSettings />;
        break;
      case SidePanelTab.LEADERBOARD:
        sidePanel = <Leaderboard />;
        break;
      case SidePanelTab.COMMIT_LOG:
        sidePanel = <CommitLog />;
        break;
      case SidePanelTab.NOTIFICATION:
        sidePanel = <Notification />;
        break;
      case SidePanelTab.MATCH:
        sidePanel = <Match />;
        break;
      case SidePanelTab.DOCS:
        sidePanel = <Docs />;
    }
    return sidePanelTab === SidePanelTab.NONE ? (
      sidePanel
    ) : (
      <div
        className={classnames(styles.SidePanel)}
        style={{
          width: `${sidePanelWidth}px`,
        }}
      >
        {sidePanel}
      </div>
    );
  }
}

export namespace SidePanel {
  export interface StateProps {
    sidePanelTab: SidePanelTab;
  }

  export interface OwnProps {
    sidePanelWidth: number;
  }

  export type Props = StateProps & OwnProps;
}
