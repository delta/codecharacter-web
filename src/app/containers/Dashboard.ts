import { DashboardActions, UserActions } from 'app/actions';
import { Dashboard } from 'app/components/Dashboard';
import { RootState } from 'app/reducers';
import { SidePanelTab } from 'app/reducers/Dashboard';
import * as DashboardInterfaces from 'app/types/Dashboard';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

const mapStateToProps = (rootState: RootState) => {
  return {
    isAuthenticationOpen: rootState.user.isAuthenticationOpen,
    isFirstLogin: rootState.user.isFirstLogin,
    isLoggedIn: rootState.user.isLoggedIn,
    isReactTourActive: rootState.dashboard.isReactTourActive,
    isUserProfileModalOpen: rootState.user.isUserProfileModalOpen,
    isWelcomeModalOpen: rootState.dashboard.isWelcomeModalOpen,
    logFile: rootState.gameLog.gameLog,
    player1DebugLog: rootState.gameLog.player1DebugLog,
    player2DebugLog: rootState.gameLog.player2DebugLog,
    sidePanelOpen: rootState.dashboard.sidePanelTab !== SidePanelTab.NONE,
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    closeWelcomeModal: () => dispatch(DashboardActions.setIsWelcomeModalOpen(false)),
    getUserDetails: () => dispatch(UserActions.getUserDetails()),
    setIsAuthenticationOpen: (isAuthenticationOpen: boolean) =>
      dispatch(UserActions.setIsAuthenticationOpen(isAuthenticationOpen)),
    toggleReactTour: () => dispatch(DashboardActions.toggleReactTour()),
  };
};

const dashboardContainer = connect<
  DashboardInterfaces.StateProps,
  DashboardInterfaces.DispatchProps,
  {}
>(
  mapStateToProps,
  mapDispatchToProps,
)(Dashboard);

export default dashboardContainer;
