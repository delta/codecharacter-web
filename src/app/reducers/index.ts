import { codeReducer } from 'app/reducers/code/Code';
import { editorReducer } from 'app/reducers/code/Editor';
import { submissionReducer } from 'app/reducers/code/Submission';
import { dashboardReducer } from 'app/reducers/Dashboard';
import { gameLogReducer } from 'app/reducers/GameLog';
import { leaderboardReducer } from 'app/reducers/Leaderboard';
import { matchesReducer } from 'app/reducers/MatchView';
import { notificationReducer } from 'app/reducers/Notification';
import { userReducer } from 'app/reducers/User';
import * as CodeInterfaces from 'app/types/code/Code';
import * as EditorInterfaces from 'app/types/code/Editor';
import * as SubmissionInterfaces from 'app/types/code/Submission';
import * as DashboardInterfaces from 'app/types/Dashboard';
import * as GameLogInterfaces from 'app/types/GameLog';
import * as LeaderboardInterfaces from 'app/types/Leaderboard';
import * as MatchInterfaces from 'app/types/MatchView';
import * as NotificationInterfaces from 'app/types/Notification';
import * as UserInterfaces from 'app/types/User';
import { routerReducer, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';

export const rootReducer = combineReducers({
  code: codeReducer,
  dashboard: dashboardReducer,
  editor: editorReducer,
  gameLog: gameLogReducer,
  leaderboard: leaderboardReducer,
  match: matchesReducer,
  notification: notificationReducer,
  router: routerReducer,
  submission: submissionReducer,
  user: userReducer,
});

export interface RootState {
  code: CodeInterfaces.CodeStoreState;
  dashboard: DashboardInterfaces.DashboardStoreState;
  editor: EditorInterfaces.EditorStoreState;
  leaderboard: LeaderboardInterfaces.LeaderboardStoreState;
  match: MatchInterfaces.MatchStoreState;
  notification: NotificationInterfaces.NotificationStoreState;
  router: RouterState;
  gameLog: GameLogInterfaces.GameLogStoreState;
  user: UserInterfaces.UserStoreState;
  submission: SubmissionInterfaces.SubmissionStoreState;
}
