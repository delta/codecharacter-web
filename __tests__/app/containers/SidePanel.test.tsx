import { DashboardActions } from 'app/actions';
import SidePanel from 'app/containers/SidePanel';
import { SidePanelTab } from 'app/reducers/Dashboard';
import { configureStore } from 'app/store';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('SidePanel Container', () => {
  const { store } = configureStore();

  it('Should render empty <div/> initially', () => {
    const wrapper = shallow(<SidePanel sidePanelWidth={350} />, {
      context: {
        store,
      },
    });
    expect(wrapper.find('.SidePanel').children()).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render EditorSettings', () => {
    store.dispatch(DashboardActions.setSidePanelTab(SidePanelTab.EDITOR_SETTINGS));
    const wrapper = shallow(<SidePanel sidePanelWidth={350} />, {
      context: {
        store,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render Leaderboard', () => {
    store.dispatch(DashboardActions.setSidePanelTab(SidePanelTab.LEADERBOARD));
    const wrapper = shallow(<SidePanel sidePanelWidth={350} />, {
      context: {
        store,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render empty <div/> after sidePanel toggle off', () => {
    store.dispatch(DashboardActions.setSidePanelTab(SidePanelTab.NONE));
    const wrapper = shallow(<SidePanel sidePanelWidth={350} />, {
      context: {
        store,
      },
    });
    expect(wrapper.find('.SidePanel').children()).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });
});
