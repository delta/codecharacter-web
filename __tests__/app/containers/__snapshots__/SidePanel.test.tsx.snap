// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePanel Container Should render EditorSettings 1`] = `
<SidePanel
  dispatch={[Function]}
  sidePanelTab="EDITOR_SETTINGS"
  sidePanelWidth={350}
/>
`;

exports[`SidePanel Container Should render Leaderboard 1`] = `
<SidePanel
  dispatch={[Function]}
  sidePanelTab="LEADERBOARD"
  sidePanelWidth={350}
/>
`;

exports[`SidePanel Container Should render empty <div/> after sidePanel toggle off 1`] = `
<SidePanel
  dispatch={[Function]}
  sidePanelTab="NONE"
  sidePanelWidth={350}
/>
`;

exports[`SidePanel Container Should render empty <div/> initially 1`] = `
<SidePanel
  dispatch={[Function]}
  sidePanelTab="NONE"
  sidePanelWidth={350}
/>
`;
